arrow-fx-coroutines/README.MD

```diff
@@ -1,561 +1,561 @@
 
-`raceN` is similar to `parMapN` except for it only returns the winner of the race, and cancels the loser as soon as the first fiber finishes the race.
+```kotlin
+internal val scheduler: ScheduledExecutorService = TODO()
+
+suspend fun sleep(duration: Duration): Unit =
+  if (duration.amount <= 0) Unit
+  else cancellable { resumeWith: (Result<Unit>) -> Unit ->
+    val cancelRef = scheduler.schedule(
+      { resumeWith(Result.success(Unit)) },
+      duration.amount,
+      duration.timeUnit
+    )
+
+    CancelToken { cancelRef.cancel(false) }
+  }
+```
 
-#### RacePair/RaceTriple/ForkAndForget
+We open a `cancellable` block and get a callback which we can invoke with either a pure value, or a `Throwable` failure. We can then schedule a task to resume the callback after the desired `Duration` has passed, and to the `cancellable` block we return the `CancelToken` that we want to execute if the resulting `suspend fun` gets cancelled.
 
-There are a couple of low level constructs available, and they should be used with care!
+So the signature of `cancellable` looks like:
 
-- racePair/raceTriple like `raceN` races 2 operations, but instead of cancelling the loser it returns the `Fiber` of the loser task
+`suspend fun <A> cancellable(cb: ((Result<A>) -> Unit) -> CancelToken): A`
 
-- ForkAndForget launches a task `fa` as a `Fiber` but doesn't connect it's cancellation system
+Where we create a `cancellable` `suspend fun` by consuming a callback of the shape `((Result<A>) -> Unit)` and returning `CancelToken`. 
 
-Both operators can safely capture the running `Fiber` and _guarantee_ not leaking using `bracketCase`.
-An example, writing `raceN` in terms of `racePair`.
+So let's take a look at what the implementation of `cancellable` looks like.
 
-```kotlin:ank
-suspend fun <A, B> raceN(ctx: CoroutineContext, fa: suspend () -> A, fb: suspend () -> B): Either<A, B> =
-  bracket(
-    acquire = { racePair(ctx, fa, fb) },
-    use = { it.bimap({ (a, _) -> a }, { (_, b) -> b }) },
-    release = {
-      it.bimap(
-        { (_, fiber) -> fiber.cancel() },
-        { (fiber, _) -> fiber.cancel() }
-      )
+```
+suspend fun <A> cancellable(cb: ((Result<A>) -> Unit) -> CancelToken): A =
+  suspendCoroutine { cont ->
+    val conn = cont.context[SuspendConnection] ?: SuspendConnection.Uncancellable
+
+    val cancellable = ForwardCancellable()
+    conn.push(cancellable.cancel())
+
+    val token = try {
+      cb(cont::resumeWith)
+    } catch (throwable: Throwable) {
+      cont.resumeWith(Result.failure(throwable.nonFatalOrThrow()))
+      CancelToken.unit
     }
-  )
+    
+    cancellable.complete(token)
+  }
 ```
 
-## And more…
+We `suspendCoroutine` to access the `Continuation`, and since we use `suspendCoroutine` here it will also `interept` and thus behave in the same safe way as Kotlin's high-level combinators.
```

Comment:
```suggestion
We `suspendCoroutine` to access the `Continuation`, and since we use `suspendCoroutine` here it will also `intercept` and thus behave in the same safe way as Kotlin's high-level combinators.
```